File: gameserver/src/main/java/l2s/gameserver/skills/effects/i_dispel_by_slot.java
package l2s.gameserver.skills.effects;

import l2s.gameserver.model.Creature;
import l2s.gameserver.model.Skill;
import l2s.gameserver.model.actor.instances.creature.Abnormal;
import l2s.gameserver.network.l2.components.SystemMsg;
import l2s.gameserver.network.l2.s2c.SystemMessagePacket;
import l2s.gameserver.skills.AbnormalType;
import l2s.gameserver.skills.EffectType;
import l2s.gameserver.templates.skill.EffectTemplate;
import l2s.gameserver.utils.EffectsComparator;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class i_dispel_by_slot extends i_abstract_effect
{
	private final AbnormalType _abnormalType;
	private final int _maxAbnormalLvl;
	private final boolean _self;

	public i_dispel_by_slot(Creature creature, Creature target, Skill skill, boolean reflected, EffectTemplate template)
	{
		super(creature, target, skill, reflected, template);
		_abnormalType = template.getParam().getEnum("abnormal_type", AbnormalType.class);
		if(_abnormalType == AbnormalType.none)
			_maxAbnormalLvl = 0;
		else
			_maxAbnormalLvl = template.getParam().getInteger("max_abnormal_level", 0);
		_self = template.getEffectType() == EffectType.i_dispel_by_slot_myself;
	}

	@Override
	public boolean checkCondition()
	{
		return _maxAbnormalLvl != 0 && super.checkCondition();
	}

	@Override
	public void instantUse()
	{
		Creature target = _self ? getEffector() : getEffected();
        List<Abnormal> effects = new ArrayList<>(target.getAbnormalList().getEffects());
		effects.sort(EffectsComparator.getInstance());
		Collections.reverse(effects);

        List<Skill> dispelledSkills = new ArrayList<>();
        for(Abnormal effect : effects)
		{
			Skill effectSkill = effect.getSkill();
			if(effectSkill == null)
				continue;
			if(effectSkill.isToggle())
				continue;
			if(effectSkill.isPassive())
				continue;
			if(effect.getAbnormalType() != _abnormalType)
				continue;
			if(_maxAbnormalLvl != -1 && effect.getAbnormalLvl() > _maxAbnormalLvl)
				continue;

			if(!effect.isOffensive())
				effect.scheduleForReApply(30_000);

			effect.exit();

			if(effect.isHidden() || dispelledSkills.contains(effectSkill))
				continue;
			dispelledSkills.add(effectSkill);
			target.sendPacket(new SystemMessagePacket(SystemMsg.THE_EFFECT_OF_S1_HAS_BEEN_REMOVED).addSkillName(effectSkill));
		}
	}
}

File: gameserver/src/main/java/l2s/gameserver/handler/effects/impl/instant/retail/i_dispel_by_slot.kt
package l2s.gameserver.handler.effects.impl.instant.retail

import l2s.gameserver.Config
import l2s.gameserver.handler.effects.impl.instant.i_abstract_effect
import l2s.gameserver.model.Creature
import l2s.gameserver.model.actor.instances.creature.Abnormal
import l2s.gameserver.model.actor.instances.player.Cubic
import l2s.gameserver.network.l2.components.SystemMsg
import l2s.gameserver.network.l2.s2c.SystemMessagePacket
import l2s.gameserver.skills.AbnormalType
import l2s.gameserver.templates.skill.EffectTemplate
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Dispel By Slot effect implementation.
 * @author Gnacik, Zoey76, Adry_85
 * @author Java-man
 */
class i_dispel_by_slot(template: EffectTemplate) : i_abstract_effect(template) {

    private val abnormalType: AbnormalType
    private val param2: Int

    init {
        val name = params.getString("i_dispel_by_slot_param1").toUpperCase()
        abnormalType = AbnormalType.valueOf(name)
        require(abnormalType != AbnormalType.NONE) {
            "Skill ${skill.id} abnormal_type should not be NONE."
        }

        param2 = params.getInteger("i_dispel_by_slot_param2")
    }

    override fun instantUse(caster: Creature, target: Creature, soulShotUsed: AtomicBoolean, reflected: Boolean, cubic: Cubic) {
        val targetCreature = target.asCreature() ?: return

        if (abnormalType == AbnormalType.TRANSFORM) {
            if (targetCreature.abnormalList.contains(AbnormalType.TRANSFORM)) {
                if (targetCreature.transformId == param2 || param2 < 0) {
                    targetCreature.transform = null
                    return
                }
            }
        }

        val abnormalList = targetCreature.abnormalList
        abnormalList
                .filter { canBeDispelled(caster, targetCreature, it) }
                .forEach {
                    removeAbnormal(caster, it)
                }
    }

    fun canBeDispelled(caster: Creature, target: Creature, abnormal: Abnormal): Boolean {
        /*if(!abnormal.isCancelable())
                 return false;*/

        val effectSkill = abnormal.skill ?: return false

        if (effectSkill.isToggle)
            return false

        if (effectSkill.isPassive)
            return false

        if (target.isSpecialAbnormal(effectSkill))
            return false

        if (!abnormal.abnormalTypeList.contains(abnormalType))
            return false

        if (param2 >= 0 && abnormal.abnormalLvl > param2) {
            return false
        }

        return true
    }

    private fun removeAbnormal(caster: Creature, abnormal: Abnormal) {
        abnormal.exit()

        if (!Config.SHOW_TARGET_PLAYER_DEBUFF_EFFECTS && !abnormal.isHidden) {
            val packet =
                    SystemMessagePacket(SystemMsg.THE_EFFECT_OF_S1_HAS_BEEN_REMOVED)
                            .addSkillName(abnormal.skill)
            caster.sendPacket(packet)
        }
    }

}